File: sprint1/go/lesson7code6.cpp
func getLeastPrimesLinear(n int) ([]int, []int) {
    lp := make([]int, n+1)
    primes := []int{}
    for i := 2; i <= n; i++ {
        if lp[i] == 0 {
            lp[i] = i
            primes = append(primes, i)
        }
        for _, p := range primes {
            x := p * i
            if p > lp[i] || x > n {
                break
            }
            lp[x] = p
        }
    }
    return primes, lp
}

File: sprint1/C#/lesson7code6.cs
public static (List<int>, List<int>) GetLeastPrimesLinear(int n)
{
    List<int> lp = Enumerable.Repeat(0, n + 1).ToList();
    List<int> primes = new List<int>();

    for (int i = 2; i <= n; i++)
    {
        if (lp[i] == 0)
        {
            lp[i] = i;
            primes.Add(i);
        }

        foreach (int p in primes)
        {
            int x = p * i;

            if (p > lp[i] || x > n)
            {
                break;
            }

            lp[x] = p;
        }
    }

    return (primes, lp);
}
